iggy/src/users/update_user.rs

```diff
@@ -0,0 +1,234 @@
+use crate::bytes_serializable::BytesSerializable;
+use crate::command::CommandPayload;
+use crate::error::Error;
+use crate::identifier::Identifier;
+use crate::models::user_status::UserStatus;
+use crate::users::{MAX_USERNAME_LENGTH, MIN_USERNAME_LENGTH};
+use crate::utils::text;
+use crate::validatable::Validatable;
+use bytes::BufMut;
+use serde::{Deserialize, Serialize};
+use std::fmt::Display;
+use std::str::{from_utf8, FromStr};
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Default)]
+pub struct UpdateUser {
+    #[serde(skip)]
+    pub user_id: Identifier,
+    pub username: Option<String>,
+    pub status: Option<UserStatus>,
+}
+
+impl CommandPayload for UpdateUser {}
+
+impl Validatable<Error> for UpdateUser {
+    fn validate(&self) -> Result<(), Error> {
+        if self.username.is_none() {
+            return Ok(());
+        }
+
+        let username = self.username.as_ref().unwrap();
+        if username.is_empty()
+            || username.len() > MAX_USERNAME_LENGTH
+            || username.len() < MIN_USERNAME_LENGTH
+        {
+            return Err(Error::InvalidUsername);
+        }
+
+        if !text::is_resource_name_valid(username) {
+            return Err(Error::InvalidUsername);
+        }
+
+        Ok(())
+    }
+}
+
+impl FromStr for UpdateUser {
+    type Err = Error;
+    fn from_str(input: &str) -> Result<Self, Self::Err> {
+        let parts = input.split('|').collect::<Vec<&str>>();
+        if parts.is_empty() || parts.len() > 3 {
+            return Err(Error::InvalidCommand);
+        }
+
+        let user_id = parts[0].parse::<Identifier>()?;
+        let username = match parts.get(1) {
+            Some(username) => match *username {
+                "" => None,
+                _ => Some(username.to_string()),
+            },
+            None => None,
+        };
+        let status = match parts.get(2) {
+            Some(status) => match *status {
+                "" => None,
+                _ => {
+                    let status = UserStatus::from_str(status)?;
+                    Some(status)
+                }
+            },
+            None => None,
+        };
+        let command = UpdateUser {
+            user_id,
```

Comment:
you can do it cleaner, someting like:
```rust
        let (user_id_str, username_str, status_str) = match parts.as_slice() {
            [user_id, username, status] => (Some(*user_id), Some(*username), Some(*status)),
            [user_id, username] => (Some(*user_id), Some(*username), None),
            [user_id] => (Some(*user_id), None, None),
            _ => return Err(Error::InvalidCommand),
        };
```

and for user_id, something like:
```rust
        let user_id = user_id_str
            .ok_or_else(|| Error::InvalidCommand)?
            .parse::<Identifier>()
            .map_err(|_| Error::InvalidCommand)?;
```

tbh it's not perfect but IMHO better (it would be better if there was context to the error)

Reply:
The only issue is that it would still catch the `""` as the valid value - in the future it shouldn't be the case, as most likely this code will be either removed or refactored as it's used only by the CLI for now.